### PR TITLE
[GPU] Shapeless convolution: add non-group specializaiton

### PR DESCRIPTION
**Pull request by echeresh**
Found some issues while looking at performance data from [MFDNN-13246](https://jira.devtools.intel.com/browse/MFDNN-13246).

List of changes:

- Added `g1` specialization - to substitute `g = 1` (one group). This results in more broad usage of block 2D based kernels
- Fixed Stream-K batch offsets (found while running the Jira shapes)
- Minor refactoring to reuse common macros for bitmask enums

### CODE REVIEW

**Review by atkassen — APPROVED** [review]

File: src/gpu/intel/jit/utils/utils.hpp [hunk 1]
```diff
@@ -1228,6 +1228,20 @@ void deserialize_from_hex(T &t, const std::string &s_hex) {
     d.pop(t);
 }
 
+#define GPU_DEFINE_BIT_MASK_ENUM_OPS(E) \
+    inline E operator&(E a, E b) { \
+        return static_cast<E>( \
+                static_cast<uint32_t>(a) & static_cast<uint32_t>(b)); \
+    } \
```
[the remark this answers is not in the document]
Comment (echeresh) [reply src/gpu/intel/jit/utils/utils.hpp:1235]: Thanks, updated.

File: src/gpu/intel/jit/v2/conv/gen_convolution.cpp [hunk 2]
```diff
@@ -135,15 +135,15 @@ class gen_convolution_t {
         kernel_desc_t _desc;
         if (debug_t::init_kernel_desc(_desc)) {
             _desc.set_defaults();
+            _desc.spec.mode = default_specialization_mode;
```
[the remark this answers is not in the document]
Comment (echeresh) [reply src/gpu/intel/jit/v2/conv/gen_convolution.cpp:138]: These are different defaults :) For example `set_defaults()` is called after parsing/deserialization and we don't want to set any extra modes. However in this particular place we want to add "default" specialization.
But I see - it's confusing now so I renamed `set_defaults()` to `set_missing()` - should be less confusing now.

File: src/gpu/intel/jit/v2/conv/planner/bench.cpp [hunk 3]
```diff
@@ -593,24 +593,25 @@ std::vector<problem_t> load_problems(const std::string &path) {
 }
 
 bench_data_t bench(const bench_manager_t &bench_mger,
-        const kernel_desc_t &kernel_desc, std::vector<bench_task_t> &tasks,
+        const kernel_desc_t &_kernel_desc, std::vector<bench_task_t> &tasks,
```
[the remark this answers is not in the document]
Comment (echeresh) [reply src/gpu/intel/jit/v2/conv/planner/bench.cpp:596]: I think this change is not fully correct: the original kernel descriptor was stored in the benchmark data but now the updated descriptor is passed which is not  intended. So we actually need the original object, and const reference is justified. Thanks for spotting that!

**Comment by echeresh** [comment]
make test
set test_scope=NIGHTLY
disable test_device_cpu
disable benchdnn_all
enable benchdnn_conv
enable benchdnn_deconv
enable benchdnn_matmul
enable arch_gpu_xe-hpc
enable arch_gpu_xe-hpg-atsm
enable arch_gpu_xe-hpg-dg2
enable arch_gpu_xe-lp
enable arch_gpu_xe-lpg
enable arch_gpu_xe-lpg+
enable arch_gpu_xe2-hpg-bmg
enable arch_gpu_xe2-lpg
enable arch_gpu_xe3-lpg

File: src/gpu/intel/jit/v2/conv/kernel_desc.hpp [hunk 4]
```diff
@@ -62,20 +62,29 @@ struct hw_desc_t {
...
+enum class specialization_mode_t : uint32_t {
+    none = 0,
     // Whether to specialize all problem values.
-    max,
+    max = 1,
     // Whether to reduce dimensions based on the problem, e.g. 3D -> 2D.
-    min_dims
+    min_dims = 2,
+    // Whether to specialize to non-group convolution (number of groups is one).
+    g1 = 4,
```
Comment (rjoursler) [inline src/gpu/intel/jit/v2/conv/kernel_desc.hpp:72]: Would it be reasonable to just generalize `min_dims` to specialize all size 1 dimensions? I don't see much downside to this as size 1 dimensions seems like a generally useful specialization.

Comment (rjoursler) [reply src/gpu/intel/jit/v2/conv/kernel_desc.hpp:72]: > If you don't have objections - I'll go ahead and rename min_dims -> _default keeping the old approach without bitmasks.

Sounds good, there is no reason to complicate the unnecessarily at this point.

> For anything in between - we can specialize the dimensions based on --spec <...> parameter.

I don't think the current controls for  `--spec` will be sufficient. The issue I expect is that we will want to conditionally specialize. In particular, my expectation is that we will have two main reasons to specialize

- Specialization enable a transformation that provides significant performance improvement
- Specialization provides minor benefit, but workloads rarely need both multiple specializations of a given strategy

I expect most cases will be well handled by just have a default strategy. but once we get into the weeds, the edge case will require a more robust mechanism for representing when a dimension should be specialized.

**Comment by echeresh** [comment]
make test
set test_scope=NIGHTLY
disable test_device_cpu
disable benchdnn_all
enable benchdnn_conv
enable benchdnn_deconv
enable arch_gpu_xe-hpc
enable arch_gpu_xe-hpg-atsm
enable arch_gpu_xe-hpg-dg2
enable arch_gpu_xe-lp
enable arch_gpu_xe-lpg
enable arch_gpu_xe-lpg+
enable arch_gpu_xe2-hpg-bmg
enable arch_gpu_xe2-lpg
enable arch_gpu_xe3-lpg

**Comment by echeresh** [comment]
make test
set test_scope=NIGHTLY
disable test_device_cpu
disable benchdnn_all
enable benchdnn_conv
enable benchdnn_deconv
enable arch_gpu_xe-hpc
enable arch_gpu_xe-hpg-atsm
enable arch_gpu_xe-hpg-dg2
enable arch_gpu_xe-lp
enable arch_gpu_xe-lpg
enable arch_gpu_xe-lpg+
enable arch_gpu_xe2-hpg-bmg
enable arch_gpu_xe2-lpg
enable arch_gpu_xe3-lpg

**Review by rjoursler — APPROVED** [review]

**Comment by echeresh** [comment]
make test
set test_scope=NIGHTLY
disable test_device_cpu
disable benchdnn_all
enable benchdnn_conv
enable benchdnn_deconv
enable benchdnn_matmul
enable arch_gpu_xe-hpc
enable arch_gpu_xe-hpg-atsm
enable arch_gpu_xe-hpg-dg2
enable arch_gpu_xe-lp
enable arch_gpu_xe-lpg
enable arch_gpu_xe-lpg+
enable arch_gpu_xe2-hpg-bmg
enable arch_gpu_xe2-lpg
enable arch_gpu_xe3-lpg
